src/comm/XPUPair.h

```diff
@@ -0,0 +1,246 @@
+  pair(const std::pair<U1, U2>& p) : first(p.first), second(p.second) {}
+
+  inline void swap(pair& p) {
+    swap(first, p.first);
+    swap(second, p, second);
+  }
+};
+
+template <typename T1, typename T2>
+bool operator==(const pair<T1, T2>& x, const pair<T1, T2>& y) {
+  return x.first == y.first && x.second == y.second;
+}
+
+template <typename T1, typename T2>
+inline bool operator<(const pair<T1, T2>& x, const pair<T1, T2>& y) {
+  return x.first < y.first || (!(y.first < x.first) && x.second < y.second);
+}
+
+template <typename T1, typename T2>
+inline bool operator!=(const pair<T1, T2>& x, const pair<T1, T2>& y) {
+  return !(x == y);
+}
+
+template <typename T1, typename T2>
+inline bool operator>(const pair<T1, T2>& x, const pair<T1, T2>& y) {
+  return y < x;
+}
+
+template <typename T1, typename T2>
+bool operator<=(const pair<T1, T2>& x, const pair<T1, T2>& y) {
+  return !(y < x);
+}
+
+template <typename T1, typename T2>
+bool operator>=(const pair<T1, T2>& x, const pair<T1, T2>& y) {
+  return !(x < y);
+}
+
+template <typename T1, typename T2>
+inline void swap(pair<T1, T2>& x, pair<T1, T2>& y) {
+  return x.swap(y);
+}
+
+template <typename T1, typename T2>
+inline pair<T1, T2> make_pair(T1 x, T2 y) {
+  return pair<T1, T2>(x, y);
+}
+
+template <size_t N, class T>
+struct tuple_element;
+
+template <size_t N, class T>
+struct tuple_element_impl {
+ private:
+  typedef typename T::tail_type Next;
+
+ public:
+  typedef typename tuple_element_impl<N - 1, Next>::type type;
+};
+
+template <class T>
+struct tuple_element_impl<0, T> {
+  typedef typename T::head_type type;
+};
+
+template <size_t N, class T>
+struct tuple_element<N, T const> {
+  using type =
+      typename std::add_const<typename tuple_element<N, T>::type>::type;
+};
+
+template <size_t N, class T>
+struct tuple_element<N, T volatile> {
+  using type =
+      typename std::add_volatile<typename tuple_element<N, T>::type>::type;
+};
+
+template <size_t N, class T>
+struct tuple_element<N, T const volatile> {
+  using type = typename std::add_cv<typename tuple_element<N, T>::type>::type;
+};
+
+template <size_t N, class T>
+struct tuple_element {
+  using type = typename tuple_element_impl<N, T>::type;
+};
+
+template <typename T1, typename T2>
+struct tuple_element<0, pair<T1, T2>> {
+  typedef T1 type;
+};
+
+template <typename T1, typename T2>
+struct tuple_element<1, pair<T1, T2>> {
+  typedef T2 type;
+};
+
+// forward declaration of tuple_size
+template <class T>
+struct tuple_size;
+
+template <class T>
+struct tuple_size<T const> : public tuple_size<T> {};
+
+template <class T>
+struct tuple_size<T volatile> : public tuple_size<T> {};
+
+template <class T>
+struct tuple_size<T const volatile> : public tuple_size<T> {};
+
+template <class T>
+struct tuple_size {
+  static const int value = 1 + tuple_size<typename T::tail_type>::value;
+};
+
+// specializations for tuple_size
+template <>
+struct tuple_size<tuple<>> {
```

Comment:
I am not sure if `tuple` and `tuple_size` are valid structs as they are defined as forward declarations.